### PR TITLE
Converted snippet preview mode buttons to radio buttons

### PR DESCRIPTION
**Pull request by hwinne**
## Summary

**_This issue has a PR in both the [javascript](https://github.com/Yoast/javascript/pull/376) and [wordpress-seo](https://github.com/Yoast/wordpress-seo/pull/13649) repo._** 

Removed the mobile and desktop mode buttons, added radio buttons. Changed the title from `Snippet Preview` to `Google search result preview` (see https://github.com/Yoast/wordpress-seo/pull/13649). Moved the buttons above the Snippet Preview.

This PR can be summarized in the following changelog entry:

* Enhances the Snippet Preview in the metabox, using radio buttons.

## Relevant technical choices:

* Changed regular mode buttons to radio buttons
* Moved mode buttons above the Snippet Preview
* Changed the Snippet Preview Title (in ).

## Test instructions

This PR can be tested by following these steps:

* The snippet preview should look like the provided screenshots down below.
    * Test on all pages the Snippet Editor is used (such as: Posts, Pages, Tags, Categories, etc...)
* The snipper preview should have the title "Google search result preview"
    * In the Metabox, under a post
    * In the Gutenberg sidebar menu.
    * In the Snippet Preview, when clicked from the Gutenberg sidebar menu.
* The mobile/desktop mode should be toggled, when the radio buttons are toggled.

## Screenshots

* Under a post:
image

* In the Yoast sidebar in Gutenberg:
image

* When the Snippet Preview is accessed from the Yoast sidebar in Gutenberg:
image

## Impact check

* This PR affects the following parts of the plugin, which may require extra testing:
   * Yoast Sidebar in the Gutenberg editor
   * Any page that uses the Snippet Preview in the Metabox

## UI changes
* [x] This PR changes the UI in the plugin. I have added the 'UI change' label to this PR.

## Quality assurance

* [x] I have tested this code to the best of my abilities
* [ ] I have added unittests to verify the code works as intended

Fixes #https://github.com/Yoast/wordpress-seo/issues/13192

### CODE REVIEW

**Comment by hwinne** [comment]
🚧 The tests should be updated.

File: packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js [hunk 1]
```diff
@@ -4,99 +4,75 @@ import styled from "styled-components";
...
-			<ScreenReaderText>
-				{ __( "Desktop preview", "yoast-components" ) }
-			</ScreenReaderText>
-		</DesktopButton>
-	</Switcher>;
+	return ( <Switcher>
+		<SwitcherTitle>{ __( "Preview as", "yoast-components" ) + ":" }</SwitcherTitle>
+		<ModeLabel>
+			<ModeRadio
+				onChange={ () => onChange( MODE_MOBILE ) }
```
Comment (afercia) [inline packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js]: Pre-existing ESLint warning: JSX props should not use arrow functions

File: packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js [hunk 2]
```diff
@@ -4,99 +4,75 @@ import styled from "styled-components";
...
+				aria-pressed={ active === MODE_MOBILE }
+				type="radio"
+				name="screen"
+				value="mobile"
+			/>
+			<span>{ __( "Mobile result", "yoast-components" ) }</span>
+		</ModeLabel>
+		<ModeLabel>
+			<ModeRadio
+				onChange={ () => onChange( MODE_DESKTOP ) }
```
Comment (afercia) [inline packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js]: Pre-existing ESLint warning: JSX props should not use arrow functions

File: packages/search-metadata-previews/src/snippet-editor/SnippetEditor.js [hunk 3]
```diff
@@ -528,6 +528,8 @@ class SnippetEditor extends React.Component {
 		return (
 			<ErrorBoundary>
 				<div>
+					<ModeSwitcher onChange={ ( newMode ) => onChange( "mode", newMode ) } active={ mode } />
```
Comment (afercia) [inline packages/search-metadata-previews/src/snippet-editor/SnippetEditor.js]: Pre-existing ESLint warning: JSX props should not use arrow functions

File: packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js [hunk 4]
```diff
@@ -4,99 +4,75 @@ import styled from "styled-components";
...
+const Switcher = styled.div`
+	display: inline-block;
+	margin-bottom: 16px;
+	vertical-align: top;
 `;
 
-const MobileButton = styled( SwitcherButton )`
-	border-radius: 3px 0 0 3px;
+const ModeLabel = styled.label`
+	font: 400 14px/24px "Open Sans",sans-serif;
```
Comment (afercia) [inline packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js]: Not sure we should use `"Open Sans",sans-serif`.  Unless I'm missing something I think we normally use "font-family: inherit" and set the other values separately.

File: packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js [hunk 5]
```diff
@@ -4,99 +4,75 @@ import styled from "styled-components";
...
+	margin-right: 18px;
```
Comment (afercia) [inline packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js]: Why `18px`? I guess this should use our 8 pixels grid thus `16px` would be a better value.
**More importantly**: this needs to be reverted for RTL.

File: packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js [hunk 6]
```diff
@@ -4,99 +4,75 @@ import styled from "styled-components";
...
+	cursor: pointer;
+	color: #3c4043;
 `;
 
-const DesktopButton = styled( SwitcherButton )`
-	border-radius: 0 3px 3px 0;
+const ModeRadio = styled.input`
+	margin-right: 10px;
```
Comment (afercia) [inline packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js]: Why `10px`? I guess this should use our 8 pixels grid thus `8px` would be a better value.
**More importantly**: this needs to be reverted for RTL.

File: packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js [hunk 7]
```diff
@@ -4,99 +4,75 @@ import styled from "styled-components";
...
-
-	&:hover, &:focus {
-		background-color: ${ colors.$color_white };
-		border: none;
-		border-bottom: 4px solid transparent;
-		border-color: ${ colors.$color_snippet_focus };
-		color: ${ colors.$color_snippet_focus };
-		box-shadow: none;
-	}
+const Switcher = styled.div`
```
Comment (afercia) [inline packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js]: This should be a `<fieldset>` element to properly group the radio buttons.

File: packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js [hunk 8]
```diff
@@ -4,99 +4,75 @@ import styled from "styled-components";
...
 `;
 
-const Switcher = styled.div`
-	display: inline-block;
-	margin-top: 10px;
-	border: 1px solid #dbdbdb;
-	border-radius: 4px;
-	background-color: #f7f7f7;
-	vertical-align: top;
+const SwitcherTitle = styled.h4`
```
Comment (afercia) [inline packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js]: This should be the `<legend>` of the `<fieldset>` element.

File: packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js [hunk 9]
```diff
@@ -4,99 +4,75 @@ import styled from "styled-components";
...
-			aria-pressed={ active === MODE_DESKTOP }
-		>
-			<SvgIcon icon="desktop" size="18px" color="currentColor" />
-			<ScreenReaderText>
-				{ __( "Desktop preview", "yoast-components" ) }
-			</ScreenReaderText>
-		</DesktopButton>
-	</Switcher>;
+	return ( <Switcher>
+		<SwitcherTitle>{ __( "Preview as", "yoast-components" ) + ":" }</SwitcherTitle>
```
Comment (afercia) [inline packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js]: I know this `+ ":"` pattern is used in a few other places but I'd reconsider it. What if other languages don't use colons? Translator wouldn't be able to remove it. I'd tend to think the colon should be part of the translatable string.

File: packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js [hunk 10]
```diff
@@ -4,99 +4,75 @@ import styled from "styled-components";
...
+		<ModeLabel>
+			<ModeRadio
+				onChange={ () => onChange( MODE_MOBILE ) }
+				checked={ active === MODE_MOBILE }
+				aria-pressed={ active === MODE_MOBILE }
```
Comment (afercia) [inline packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js]: Please remove `aria-pressed`: it was OK for the previous buttons but it's not appropriate for radio buttons.

File: packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js [hunk 11]
```diff
@@ -4,99 +4,75 @@ import styled from "styled-components";
...
+				name="screen"
+				value="mobile"
+			/>
+			<span>{ __( "Mobile result", "yoast-components" ) }</span>
+		</ModeLabel>
+		<ModeLabel>
+			<ModeRadio
+				onChange={ () => onChange( MODE_DESKTOP ) }
+				checked={ active === MODE_DESKTOP }
+				aria-pressed={ active === MODE_DESKTOP }
```
Comment (afercia) [inline packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js]: Please remove `aria-pressed`: it was OK for the previous buttons but it's not appropriate for radio buttons.

File: packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js [hunk 12]
```diff
@@ -4,99 +4,75 @@ import styled from "styled-components";
...
-		>
-			<SvgIcon icon="desktop" size="18px" color="currentColor" />
-			<ScreenReaderText>
-				{ __( "Desktop preview", "yoast-components" ) }
-			</ScreenReaderText>
-		</DesktopButton>
-	</Switcher>;
+	return ( <Switcher>
+		<SwitcherTitle>{ __( "Preview as", "yoast-components" ) + ":" }</SwitcherTitle>
+		<ModeLabel>
```
Comment (afercia) [inline packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js]: In other implementations we don't use wrapping labels. Instead, we use separate label and input, associated with `for` and `id` attributes. See for example the Configuration wizard. That's the recommended way to go as some assistive technologies don't work well with wrapping labels.

Comment (afercia) [reply packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js]: Also, I'd consider to use the existing `Label` component.

File: packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js [hunk 13]
```diff
@@ -4,99 +4,75 @@ import styled from "styled-components";
...
+			<ModeRadio
```
Comment (afercia) [inline packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js]: I'd consider to use the existing `Input` component (see the Configuration wizard for usage example)

File: packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js [hunk 14]
```diff
@@ -4,99 +4,75 @@ import styled from "styled-components";
...
+				checked={ active === MODE_MOBILE }
+				aria-pressed={ active === MODE_MOBILE }
+				type="radio"
+				name="screen"
+				value="mobile"
+			/>
+			<span>{ __( "Mobile result", "yoast-components" ) }</span>
+		</ModeLabel>
+		<ModeLabel>
+			<ModeRadio
```
Comment (afercia) [inline packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js]: Same as above: consider to use the existing `Label` and `Input` components.

File: packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js [hunk 15]
```diff
@@ -4,99 +4,75 @@ import styled from "styled-components";
...
+				onChange={ () => onChange( MODE_DESKTOP ) }
+				checked={ active === MODE_DESKTOP }
+				aria-pressed={ active === MODE_DESKTOP }
+				type="radio"
+				name="screen"
+				value="desktop"
+			/>
+			<span>{ __( "Desktop result", "yoast-components" ) }</span>
```
Comment (afercia) [inline packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js]: Unless I'm missing something and it actually serves some purpose, please remove this `<span>` element.

File: packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js [hunk 16]
```diff
@@ -4,99 +4,75 @@ import styled from "styled-components";
...
+		<ModeLabel>
+			<ModeRadio
+				onChange={ () => onChange( MODE_MOBILE ) }
+				checked={ active === MODE_MOBILE }
+				aria-pressed={ active === MODE_MOBILE }
+				type="radio"
+				name="screen"
+				value="mobile"
+			/>
+			<span>{ __( "Mobile result", "yoast-components" ) }</span>
```
Comment (afercia) [inline packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js]: Unless I'm missing something and it actually serves some purpose, please remove this `<span>` element.

File: packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js [hunk 17]
```diff
@@ -4,99 +4,75 @@ import styled from "styled-components";
...
+	vertical-align: top;
 `;
 
-const MobileButton = styled( SwitcherButton )`
-	border-radius: 3px 0 0 3px;
+const ModeLabel = styled.label`
+	font: 400 14px/24px "Open Sans",sans-serif;
+	margin-right: 18px;
+	cursor: pointer;
+	color: #3c4043;
```
Comment (afercia) [inline packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js]: This color is a Google color, only used for the snippet URL and the description. We shouldn't use it for our UI.

File: packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js [hunk 18]
```diff
@@ -4,99 +4,75 @@ import styled from "styled-components";
...
-const Switcher = styled.div`
-	display: inline-block;
-	margin-top: 10px;
-	border: 1px solid #dbdbdb;
-	border-radius: 4px;
-	background-color: #f7f7f7;
-	vertical-align: top;
+const SwitcherTitle = styled.h4`
+	margin: .5em 0;
+	color: #3c4043;
```
Comment (afercia) [inline packages/search-metadata-previews/src/snippet-editor/ModeSwitcher.js]: This color is a Google color, only used for the snippet URL and the description. We shouldn't use it for our UI.

**Comment by afercia** [comment]
CR 🚧 

Some considerations to address, see comments. Most notably:
- styles for RTL 
- better semantics: group the radio buttons with fieldset + legend

**Comment by afercia** [comment]
Changes in the last commit:

- used the existing Label and Input components
- modified Label and Input to get the className passed by styled-components
- please double check the change to Label `optionalAttributes`: may need adjustments 
- used a fieldset + legend elements to group the radio buttons
- added styles for RTL

**Comment by afercia** [comment]
Note: RTL testing in the standalone demo appears to be broken for some reason.

Thus RTL needs to be tested directly in the plugin. Screenshots:

LTR:

Screenshot 2019-10-28 at 16 15 14

RTL:

Screenshot 2019-10-28 at 16 33 17

**Comment by afercia** [comment]
Note: very recent CSS changes in Gutenberg that were merged in WordPress core as part of the RC 3 broke the radio buttons layout within the meta boxes area. The radio buttons "blue dot" is now completely misplaced. See:

https://github.com/WordPress/gutenberg/issues/18181
https://core.trac.wordpress.org/ticket/48466

This should be fixed upstream, hopefully as soon as possible because it's a regression in WP trunk.

**Comment by igorschoester** [comment]
Acceptance 🚧 

One error found:
When using the sidebar, the Google preview modal closes when you click on a radio button.

**Comment by hwinne** [comment]
As Igor mentioned, the modal closed when clicking the label of the radio button. I added `preventDefault`, so the label does not execute its default actions. @afercia could you please confirm this is the correct way of doing it?

**Comment by IreneStr** [comment]
Acceptance 👍

**Comment by afercia** [comment]
> I don't know why the modal closes if you click the label, but this is not the right way to fix it I think.

Yeah, this happened (only in Chrome) because of the duplicate IDs and I guess the Gutenmodal "click outside" being triggered. Didn't happen on Firefox.

**Pull request merged**